Userland/Libraries/LibJS/Runtime/VM.cpp

```diff
@@ -137,294 +137,334 @@ Symbol* VM::get_global_symbol(const String& description)
+    // And 8.3.5 Runtime Semantics: NamedEvaluation, https://tc39.es/ecma262/#sec-runtime-semantics-namedevaluation
+    if (is<FunctionExpression>(expression)) {
+        auto& function = static_cast<FunctionExpression const&>(expression);
+        if (!function.has_name()) {
+            return function.instantiate_ordinary_function_expression(interpreter(), global_object, name);
         }
-    }
-
-    if (specific_scope && possible_match.has_value()) {
-        if (!first_assignment && possible_match.value().declaration_kind == DeclarationKind::Const) {
-            throw_exception<TypeError>(global_object, ErrorType::InvalidAssignToConst);
-            return;
+    } else if (is<ClassExpression>(expression)) {
+        auto& class_expression = static_cast<ClassExpression const&>(expression);
+        if (!class_expression.has_name()) {
+            return TRY(class_expression.class_definition_evaluation(interpreter(), global_object, {}, name));
         }
-
-        specific_scope->put_into_environment(name, { value, possible_match.value().declaration_kind });
-        return;
-    }
-
-    if (specific_scope) {
-        specific_scope->put_into_environment(name, { value, DeclarationKind::Var });
-        return;
     }
 
-    global_object.set(name, value, Object::ShouldThrowExceptions::Yes);
+    auto value = expression.execute(interpreter(), global_object);
+    if (auto* thrown_exception = exception())
+        return JS::throw_completion(thrown_exception->value());
+    return value;
 }
 
-void VM::assign(const FlyString& target, Value value, GlobalObject& global_object, bool first_assignment, Environment* specific_scope)
+// 13.15.5.2 Runtime Semantics: DestructuringAssignmentEvaluation, https://tc39.es/ecma262/#sec-runtime-semantics-destructuringassignmentevaluation
+ThrowCompletionOr<void> VM::destructuring_assignment_evaluation(NonnullRefPtr<BindingPattern> const& target, Value value, GlobalObject& global_object)
 {
-    set_variable(target, move(value), global_object, first_assignment, specific_scope);
+    // Note: DestructuringAssignmentEvaluation is just like BindingInitialization without an environment
+    //       And it allows member expressions. We thus trust the parser to disallow member expressions
+    //       in any non assignment binding and just call BindingInitialization with a nullptr environment
+    return binding_initialization(target, value, nullptr, global_object);
 }
 
-void VM::assign(const Variant<NonnullRefPtr<Identifier>, NonnullRefPtr<BindingPattern>>& target, Value value, GlobalObject& global_object, bool first_assignment, Environment* specific_scope)
+// 7.3.25 CopyDataProperties ( target, source, excludedItems ), https://tc39.es/ecma262/#sec-copydataproperties
+ThrowCompletionOr<Object*> VM::copy_data_properties(Object& rest_object, Object const& source, HashTable<PropertyName, PropertyNameTraits> const& seen_names, GlobalObject& global_object)
```

Comment:
Not sure why this is part of the VM, belongs to `// 7.3 Operations on Objects` in Object.cpp/h